File: benchmark/synduce/constraints/sortedlist/interval_intersection.f
Inductive List = elt {Int, Int} | cons {Int, Int, List};

head = \xs: List.
  match xs with
    elt {a, b} -> {a, b}
  | cons {a, b, t} -> {a, b}
  end;

sorted = fix (
  \f: List -> Bool. \xs: List.
  match xs with
    elt {a, b} -> < a b
  | cons {a, b, t} ->
    and (< a b) (and (< a (head t).1) (f t))
  end
);

inter = \a: Int. \b: Int. fix (
  \f: List -> Bool. \xs: List.
  match xs with
    elt {c, d} -> and (not (< b c)) (not (> a d))
  | cons {c, d, t} -> or (f t) (and (not (< b c)) (not (> a d)))
  end
);

spec = \xs: List. (fix (
  \f: List -> {Bool, Int, Int}. \xs: List.
  match xs with
    elt {a, b} -> {false, a, b}
  | cons {a, b, t} ->
    let res = f t in
      {or res.1 (inter a b t), a, b}
  end
) xs).1;

target = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    elt _ -> xs
  | cons {a, b, t} -> cons {a, b, f t}
  end
);

main = \xs: List. if sorted xs then spec (target xs) else false;

File: ml-benchmark/synduce/constraints/sortedlist/interval_intersection.ml
type lst = Elt of int * int | Cons of int * int * lst

let head xs =
  match xs with
  | Elt (a, b) -> (a, b)
  | Cons (a, b, t) -> (a, b)

let rec sorted xs =
  match xs with
  | Elt (a, b) -> a < b
  | Cons (a, b, t) ->
      let h = head t in
      match h with
      | (a2, b2) -> (a < b) && ((a < a2) && (sorted t))

let inter a b =
  let rec f xs =
    match xs with
    | Elt (c, d) -> (not (b < c)) && (not (a > d))
    | Cons (c, d, t) -> (f t) || ((not (b < c)) && (not (a > d)))
  in
  f

let spec xs =
  let rec f xs =
    match xs with
    | Elt (a, b) -> (false, a, b)
    | Cons (a, b, t) ->
        let res = f t in
        match res with
        | (r1, r2, r3) -> ((r1 || (inter a b t)), a, b)
  in
  let r = f xs in
  match r with
  | (r1, r2, r3) -> r1

val target: lst -> lst compress
let rec target xs =
  match xs with
  | Elt (_, _) -> xs
  | Cons (a, b, t) -> Cons (a, b, (target t))

let program xs =
  if sorted xs then spec (target xs) else false